go-controller/pkg/node/node_smartnic.go

```diff
@@ -0,0 +1,203 @@
+			if !util.PodWantsNetwork(pod) || pod.Status.Phase == kapi.PodRunning {
+				retryPods.Delete(pod.UID)
+				return
+			}
+			_, retry := retryPods.Load(pod.UID)
+			if util.PodScheduled(pod) && retry {
+				if n.name != pod.Spec.NodeName {
+					retryPods.Delete(pod.UID)
+					return
+				}
+				vfRepName, err := n.getVfRepName(pod)
+				if err != nil {
+					klog.Infof("Failed to get rep name, %s. retrying", err)
+					retryPods.Store(pod.UID, true)
+					return
+				}
+				podInterfaceInfo, err := cni.PodAnnotation2PodInfo(pod.Annotations, true)
+				if err != nil {
+					retryPods.Store(pod.UID, true)
+					return
+				}
+				err = n.addRepPort(pod, vfRepName, podInterfaceInfo)
+				if err != nil {
+					klog.Infof("Failed to add rep port, %s. retrying", err)
+					retryPods.Store(pod.UID, true)
+				} else {
+					servedPods.Store(pod.UID, true)
+					retryPods.Delete(pod.UID)
+				}
+			}
+		},
+		DeleteFunc: func(obj interface{}) {
+			pod := obj.(*kapi.Pod)
+			klog.Infof("DeleteFunc for POD: %s/%s", pod.ObjectMeta.GetNamespace(), pod.ObjectMeta.GetName())
+			if _, ok := servedPods.Load(pod.UID); !ok {
+				return
+			}
+			servedPods.Delete(pod.UID)
+			retryPods.Delete(pod.UID)
+			vfRepName, err := n.getVfRepName(pod)
+			if err != nil {
+				klog.Errorf("Failed to get VF Representor Name from Pod: %s. Representor port may have been deleted.", err)
+				return
+			}
+			err = n.delRepPort(vfRepName)
+			if err != nil {
+				klog.Infof("Failed to delete VF representor %s. %s", vfRepName, err)
+			}
+		},
+	}, nil)
+}
+
+// getVfRepName returns the VF's representor of the VF assigned to the pod
+func (n *OvnNode) getVfRepName(pod *kapi.Pod) (string, error) {
+	smartNicCD := util.SmartNICConnectionDetails{}
+	if err := smartNicCD.FromPodAnnotation(pod.Annotations); err != nil {
+		return "", fmt.Errorf("failed to get smart-nic annotation. %v", err)
+	}
+	return util.GetSriovnetOps().GetVfRepresentorSmartNIC(smartNicCD.PfId, smartNicCD.VfId)
+}
+
+// addRepPort adds the representor of the VF to the ovs bridge
+func (n *OvnNode) addRepPort(pod *kapi.Pod, vfRepName string, ifInfo *cni.PodInterfaceInfo) error {
+	klog.Infof("addRepPort: %s", vfRepName)
+	smartNicCD := util.SmartNICConnectionDetails{}
+	if err := smartNicCD.FromPodAnnotation(pod.Annotations); err != nil {
+		return fmt.Errorf("failed to get smart-nic annotation. %v", err)
+	}
+
+	err := cni.ConfigureOVS(context.TODO(), pod.Namespace, pod.Name, vfRepName, ifInfo, smartNicCD.SandboxId)
+	if err != nil {
+		return err
+	}
+	klog.Infof("Port %s added to bridge br-int", vfRepName)
+
+	link, err := util.GetNetLinkOps().LinkByName(vfRepName)
+	if err != nil {
+		// Note(adrianc): we are lenient with cleanup in this method as pod is going to be retried anyway.
+		_ = n.delRepPort(vfRepName)
+		return fmt.Errorf("failed to get link device for interface %s", vfRepName)
+	}
+
+	klog.Infof("addRepPort: set link mtu %s", vfRepName)
```

Comment:
this and the link up logs don't seem very helpful. Just log the error returned by this function.

Reply:
removed